Improve GG::Font::DetermineLines()

LGM-Doyle opened the pull request:
This PR is a collection of improvements to `GG::Font` to improve its
performance.

Previously `Font::DetermineLines()` consumed 25% of the processing time when
the UI was stalling freeorion.  Within `DetermineLines()` most of the time was
spent parsing the XML strings into TextElements used by Font to break the
string into known style tags, unknown style tags and text. 

The PR breaks out the XML to TextElements computation into
`ComputeTextElements()`.  It then removes gotcha variants of
`DetermineLines()` so that Font can be used efficiently.  It short-circuits
logic for common operations like parsing an empty string.

Repeated use of `ComputeTextElements()` in tight loops is still a
performance hazard. I was unable to improve the usage of `boost::xpressive`
as much as I had hoped.

LGM-Doyle committed:
Add TextControl constructor that uses TextElements to improve speed.

LGM-Doyle committed:
Make TextControl use member variable instead of member accessor function.

LGM-Doyle committed:
Refactor TextControl::RecomputeLineData from SetText and ChangeTemplatedText.

The code to recompute the line data was repeated in two locations.

LGM-Doyle committed:
Remove unecessary AdjustMiniumumSize from TextControl::RecomputeLineData.

LGM-Doyle committed:
Remove unecessary AdjustMinimumSize() from TextControl::SizeMove().

geoffthemedio reviewed (dismissed):
I don't understand what most of this is doing, so can't really meaningfully review that aspect...

LGM-Doyle committed:
From review #989: Prevent short text in TextControl::SetText(text, text_element)

This prevents setting a shorter string than expected by the paired
TextControl when using the fast TextControl::SetText() variant.

LGM-Doyle committed:
Comment change for Review #989 moved to new PR.

LGM-Doyle committed:
Add copy constructor to TextControl.

LGM-Doyle committed:
Add asignment operator to TextControl.

LGM-Doyle committed:
Change ComputeTextElements to ExpensiveParseFromTextToTextElements in comments.

For review #989.
